hortonworks/streamline
# STREAMLINE-512 Backend: apply Topology's namespace to "Topology Action" and "Topology Metrics"

NOTE: This is on top of STREAMLINE-501.

There're some FIXMEs on changeset: I left FIXME which needs some discussions. This patch itself is working well indeed, but may need to fix some places (exceptions, etc.).

Since UI is not ready, we need to handle some manual steps to make this work. Note that Storm and Ambari Metrics must be installed to Ambari Cluster. I guess UI is also not ready for time-series DB query yet, so you may skip checking Ambari Metrics (via set `timeSeriesDB` to null)

Manual test steps
1. create Cluster entity -- via REST API
1. run `cluster import from Ambari` to newly created Cluster -- via REST API
1. create Namespace entity (with setting streamingEngine and timeSeriesDB) -- via REST API
1. associate STORM to newly created Namespace -- via REST API
1. associate AMBARI_METRICS to newly created Namespace -- via REST API
1. create and configure Topology -- via UI
1. get created topology and add namespace ID to entity, and put -- via REST API
1. run Topology -- via UI
1. open first page -- via UI or request list topologies with metrics -- via REST API

I can share Postman (chrome extension) exported file for manual test, so please let me know when you need it.

Btw, there're some issues to discuss:

> How to handle multiple streaming engine & time-series DB

I already brought this issue earlier this week, and proposed refactoring StreamCatalogService and some modules (including patch #329 ). Please note that this patch doesn't rely on proposed patch, but just handles 'STORM' as streaming engine and 'AMBARI_METRICS' as time-series DB. 
I'm open to hear any ideas on flexible initialization of multiple streaming engine & time-series DB, so if you have any ideas other than my proposed idea, please share so that we can discuss. Also please expose your opinion about my proposed idea.

> When to invalidate TopologyActions / TopologyMetrics instance

This patch invalidates relevant TopologyActions / TopologyMetrics instance when namespace is removed or updated, or namespace-cluster mapping is changed. But there's a chance that Service, or ServiceConfiguration, or Component are changed though the chance is small.
Do we want to invalidate for all the cases, or having time-expiring cache, or OK as it is?
